helix-core/src/history.rs

```diff
@@ -23,72 +64,259 @@ impl Default for History {
+            TimePeriod(d) => self.jump_duration_backward(d),
+        }
+    }
+
+    pub fn later(&mut self, uk: UndoKind) -> Vec<Transaction> {
+        use UndoKind::*;
+        match uk {
+            Steps(n) => self.jump_forward(n),
+            TimePeriod(d) => self.jump_duration_forward(d),
+        }
+    }
+}
+
+#[derive(Debug, PartialEq)]
+pub enum UndoKind {
+    Steps(usize),
+    TimePeriod(std::time::Duration),
+}
+
+const TIME_UNITS: &[&str] = &["seconds", "minutes", "hours", "days"];
+
+fn find_time_unit(s: &str) -> Result<&'static str, String> {
+    for unit in TIME_UNITS {
+        if unit.starts_with(&s.to_lowercase()) {
+            return Ok(unit);
+        }
+    }
+    Err(format!("incorrect time unit: {}", s))
+}
+
+static DURATION_VALIDATION_REGEX: Lazy<Regex> =
+    Lazy::new(|| Regex::new(r"^(?:\d+\s*[a-zA-Z]+\s*)+$").unwrap());
+
+static NUMBER_UNIT_REGEX: Lazy<Regex> = Lazy::new(|| Regex::new(r"(\d+)\s*([a-zA-Z]+)").unwrap());
+
+fn parse_human_duration(s: &str) -> Result<Duration, String> {
+    if !DURATION_VALIDATION_REGEX.is_match(s) {
+        return Err("duration should be composed \
+        of positive integers followed by time units"
+            .to_string());
+    }
+
+    let mut units = std::collections::HashMap::new();
+    for cap in NUMBER_UNIT_REGEX.captures_iter(s) {
+        let (n_str, short_unit) = (&cap[1], &cap[2]);
+
+        let n = match n_str.parse::<u64>() {
+            Err(_) => return Err(format!("integer too large: {}", n_str)),
+            Ok(n) => n,
+        };
+
+        let unit = find_time_unit(short_unit)?;
+        if units.contains_key(unit) {
+            return Err(format!("{} specified more than once", unit));
+        }
+        units.insert(unit, n);
+    }
 
-        // for now, simply pick the latest child (linear undo / redo)
-        if let Some((index, transaction)) = current_revision.children.last() {
-            self.cursor = *index;
+    let units_in_seconds = [
+        Some(*units.get("seconds").unwrap_or(&0)),
+        units.get("minutes").unwrap_or(&0).checked_mul(60),
+        units.get("hours").unwrap_or(&0).checked_mul(60 * 60),
+        units.get("days").unwrap_or(&0).checked_mul(60 * 60 * 24),
+    ];
+    // A checked sum.
+    let seconds = units_in_seconds
+        .iter()
+        .try_fold(0u64, |acc, x| x.and_then(|x| acc.checked_add(x)));
 
-            return Some(&transaction);
+    match seconds {
+        Some(seconds) => Ok(Duration::new(seconds, 0)),
+        None => Err("duration too large".to_string()),
+    }
+}
+
+impl std::str::FromStr for UndoKind {
+    type Err = String;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        let s = s.trim();
+        if s.is_empty() {
+            return Ok(Self::Steps(1usize));
         }
-        None
+        if let Ok(n) = s.parse::<usize>() {
+            return Ok(UndoKind::Steps(n));
+        }
+        let d = parse_human_duration(s)?;
+        Ok(Self::TimePeriod(d))
```

Comment:
Will this be better? And I think we can do some type elision.
```rust
        if s.is_empty() {
            Ok(Self::Steps(1))
        } else if let Ok(n) = s.parse() {
            Ok(UndoKind::Steps(n))
        } else {
            Ok(UndoKind::TimePeriod(parse_human_duration(s)?))
        }
```

Reply:
Sure, this looks better. I can rewrite my version.